flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java

```diff
@@ -1055,46 +1058,186 @@ private static Function instantiateHiveFunction(ObjectPath functionPath, HiveCat
 		);
 	}
 
+	private boolean isTablePartitioned(Table hiveTable) {
+		return hiveTable.getPartitionKeysSize() != 0;
+	}
+
 	// ------ stats ------
 
 	@Override
 	public void alterTableStatistics(ObjectPath tablePath, CatalogTableStatistics tableStatistics, boolean ignoreIfNotExists) throws TableNotExistException, CatalogException {
-
+		try {
+			Table hiveTable = getHiveTable(tablePath);
+			// Set table stats
+			if (needUpdateStatistics(hiveTable.getParameters(), tableStatistics)) {
+				updateStatisticsParameters(tableStatistics, hiveTable.getParameters());
+				client.alter_table(tablePath.getDatabaseName(), tablePath.getObjectName(), hiveTable);
+			}
+		} catch (TableNotExistException e) {
+			if (!ignoreIfNotExists) {
+				throw e;
+			}
+		} catch (TException e) {
+			throw new CatalogException(String.format("Failed to alter table stats of table %s", tablePath.getFullName()), e);
+		}
 	}
 
 	@Override
 	public void alterTableColumnStatistics(ObjectPath tablePath, CatalogColumnStatistics columnStatistics, boolean ignoreIfNotExists) throws TableNotExistException, CatalogException {
+		try {
+			Table hiveTable = getHiveTable(tablePath);
+			// Set table column stats. This only works for non-partitioned tables.
+			if (!isTablePartitioned(hiveTable)) {
+				client.updateTableColumnStatistics(HiveCatalogUtil.createColumnStats(hiveTable, columnStatistics.getColumnStatisticsData()));
+			}
+		} catch (TableNotExistException e) {
+			if (!ignoreIfNotExists) {
+				throw e;
+			}
+		} catch (TException e) {
+			throw new CatalogException(String.format("Failed to alter table column stats of table %s", tablePath.getFullName()), e);
+		}
+	}
 
+	private static boolean needUpdateStatistics(Map<String, String> oldParameters, CatalogTableStatistics statistics) {
+		String oldRowCount = oldParameters.getOrDefault(StatsSetupConst.ROW_COUNT, "0");
+		String oldTotalSize = oldParameters.getOrDefault(StatsSetupConst.TOTAL_SIZE, "0");
+		String oldNumFiles = oldParameters.getOrDefault(StatsSetupConst.NUM_FILES, "0");
+		String oldRawDataSize = oldParameters.getOrDefault(StatsSetupConst.RAW_DATA_SIZE, "0");
+		return statistics.getRowCount() != Long.parseLong(oldRowCount) || statistics.getTotalSize() != Long.parseLong(oldTotalSize)
+				|| statistics.getFileCount() != Integer.parseInt(oldNumFiles) || statistics.getRawDataSize() != Long.parseLong(oldRawDataSize);
+	}
+
+	private static void updateStatisticsParameters(CatalogTableStatistics tableStatistics, Map<String, String> parameters) {
```

Comment:
Nit: maybe we can switch the order of the two arguments to be consistent with needUpdateStatistics() method.